pyls/plugins/importmagic_lint.py

```diff
@@ -0,0 +1,264 @@
+
+
+log = logging.getLogger(__name__)
+
+SOURCE = 'importmagic'
+ADD_IMPORT_COMMAND = 'importmagic.addimport'
+REMOVE_IMPORT_COMMAND = 'importmagic.removeimport'
+MAX_COMMANDS = 4
+UNRES_RE = re.compile(r"Unresolved import '(?P<unresolved>[\w.]+)'")
+UNREF_RE = re.compile(r"Unreferenced import '(?P<unreferenced>[\w.]+)'")
+
+_index_cache = {}
+
+
+def _get_index(sys_path):
+    """Build index of symbols from python modules.
+    Cache the index so we don't build it multiple times unnecessarily.
+    """
+    key = tuple(sys_path)
+    if key not in _index_cache:
+        log.info("Started building importmagic index")
+        index = importmagic.SymbolIndex()
+        # The build tend to be noisy
+        index.build_index(paths=sys_path)
+        _index_cache[key] = index
+        log.info("Finished building importmagic index")
+    return _index_cache[key]
+
+
+def _get_imports_list(source, index=None):
+    """Get modules, functions and variables that are imported.
+    """
+    if index is None:
+        index = importmagic.SymbolIndex()
+    imports = importmagic.Imports(index, source)
+    imported = [i.name for i in list(imports._imports)]
+    # Go over from imports
+    for from_import in list(imports._imports_from.values()):
+        imported.extend([i.name for i in list(from_import)])
+    return imported
+
+
+@hookimpl
+def pyls_commands():
+    return [ADD_IMPORT_COMMAND, REMOVE_IMPORT_COMMAND]
+
+
+@hookimpl
+def pyls_lint(document):
+    """Build a diagnostics of unresolved and unreferenced symbols.
+    Every entry follows this format:
+        {
+            'source': 'importmagic',
+            'range': {
+                'start': {
+                    'line': start_line,
+                    'character': start_column,
+                },
+                'end': {
+                    'line': end_line,
+                    'character': end_column,
+                },
+            },
+            'message': message_to_be_displayed,
+            'severity': sevirity_level,
+        }
+
+    Args:
+        document: The document to be linted.
+    Returns:
+        A list of dictionaries.
+    """
+    scope = importmagic.Scope.from_source(document.source)
+    unresolved, unreferenced = scope.find_unresolved_and_unreferenced_symbols()
+
+    diagnostics = []
+
+    # Annoyingly, we only get the text of an unresolved import, so we'll look for it ourselves
+    for unres in unresolved:
+        for line_no, line in enumerate(document.lines):
+            pos = line.find(unres)
+            if pos < 0:
+                continue
+
+            diagnostics.append({
+                'source': SOURCE,
+                'range': {
+                    'start': {'line': line_no, 'character': pos},
+                    'end': {'line': line_no, 'character': pos + len(unres)}
+                },
+                'message': "Unresolved import '%s'" % unres,
+                'severity': lsp.DiagnosticSeverity.Hint,
```

Comment:
Should this not be a warning / error? Or does it overlap with e.g. pyflakes results

Reply:
This is a hint because we actually don't know if this is a module to be imported or just a non defined variable, so better not to trigger a false positive. The unreferenced symbols are flagged with Warning because we know in advance if the symbol is either an import or a varialbe/function and the error should be consistent.